Optimisations for WorkSync

chrisvest opened the pull request:
This is the first part of #7822 that is immediately useful without adding too much complexity.

@thobe please take a look.

chrisvest committed:
More aggressive multithreaded testing of WorkSync

chrisvest committed:
Make WorkSync implement its own locking

The `WorkSync` turns a parallel workload, that does not actually allow concurrency, into a single-threaded batch workload.
It does so by piling work up in a (wait-free) stack, and then taking a lock, and atomically grabbing the pent-up stack of work and applying it in a batch.
This locking was previously implemented with a `j.u.c.l.ReentrantLock` for its `tryLock` behaviour.
Unfortunuately, the `ReentrantLock` has a wait-set to maintain, which is implemented as a concurrent linked queue.
The `tryLock` method furthermore causes allocation every time it is called.
Thus the ReentrantLock adds a significant allocation and CPU overhead.

Since each thread that contends for the `WorkSync` lock is already allocating a `WorkUnit` for the stack, we can actually make this stack double as a wait-set.
Then we pull the locking into the `WorkSync` implementation itself.
This allows us to do `tryLock`, and indeed any kind of locking, without any additional allocation.
The `mustNotApplyWorkInParallelUnderStress` that previously took ~3.5 seconds to run on my machine now takes about 1 second.
Furthermore, this setup will allow us to make the `WorkUnit` implement `j.u.c.Future`, such that we can add a `WorkSync.applyAsync()` method that will return a future.
That part of the work is still pending.

chrisvest committed:
Clarify onSpinWait todos in WorkSync

chrisvest commented:
Build was green but has failed to notify Github, which is why there's no green check-mark.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java:
```diff
@@ -158,29 +191,127 @@ private W combine( WorkUnit<Material,W> batch )
+        static final int STATE_QUEUED = 0;
+        static final int STATE_PARKED = 1;
+        static final int STATE_DONE = 2;
+        static final int STATE_CANCELLED = 3;
```
This state is never `set`.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java:
```diff
@@ -158,29 +191,127 @@ private W combine( WorkUnit<Material,W> batch )
+        boolean isCancelled()
+        {
+            return get() == STATE_CANCELLED;
+        }
```
Since `STATE_CANCELLED` is never set, this can never be true.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java:
```diff
@@ -158,29 +191,127 @@ private W combine( WorkUnit<Material,W> batch )
+            {
+                batch = batch.next;
+            }
+            while ( batch.isCancelled() );
```
Since `isCancelled()` can never be true, this loop is superfluous.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java:
```diff
@@ -103,14 +104,45 @@ public void apply( W work )
+            return true;
+        }
+
+        // Did not get the lock, spend some time until our work has aither been completed,
```
spelling: `s/aither/either/`

thobe reviewed (commented):
There are some changes I quite liked from the previous PR that didn't make it in here:
- Exception handling for when the jobs fail. It seems like a very sensible thing to do, and throwing ExecutionException for encapsulating that exception also seemed perfectly fine.
- Breaking up `apply()` into `tryDoWork` and `checkFailure` - I thought that was an elegant split.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java:
```diff
@@ -58,9 +59,9 @@
+        this.stackEnd = new WorkUnit<>( null, null, null );
         this.stack = new AtomicReference<>( stackEnd );
-        this.lock = new ReentrantLock();
+        this.lock = new AtomicReference<Thread>();
```
explicit type argument not needed.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java:
```diff
@@ -71,7 +72,7 @@ public WorkSync( Material material )
     {
         // Schedule our work on the stack.
-        WorkUnit<Material,W> unit = new WorkUnit<>( work );
+        WorkUnit<Material,W> unit = new WorkUnit<Material,W>( work, Thread.currentThread(), stackEnd );
```
explicit type arguments not needed.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java:
```diff
@@ -158,29 +191,127 @@ private W combine( WorkUnit<Material,W> batch )
+        // know that it should wait and complete its task. In fact, it may trail our done-marking
+        // process and unpark threads in parallel.
+        batch.setAsSuccessor();
+        WorkUnit<Material,W> successor = batch;
```
Local variable `successor` not used.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java:
```diff
@@ -158,29 +191,127 @@ private W combine( WorkUnit<Material,W> batch )
+            }
+        }
+
+        private void wakeUpAsSuccessor()
```
I'd like it if this method had the same type of comments as other places about why it spins waiting for things.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java:
```diff
@@ -158,29 +191,127 @@ private W combine( WorkUnit<Material,W> batch )
+        // the other blocked threads - before we begin marking the units as done.
+        // This way, if the successor wakes up before we finish marking the units as done, it will
+        // know that it should wait and complete its task. In fact, it may trail our done-marking
+        // process and unpark threads in parallel.
```
This whole "successor" business was not included in #7822, but strangely this comment was.
Digging through the commits, the "successor" logic was introduced in _Make WorkSync implement its own locking_, and then removed in _Optimisations and asynchronous apply method for WorkSync_.
Perhaps those same optimizations could be applied here, because the "successor" stuff is quite complex (quite possibly overly complex), and at best under documented (more comments please!), at worst unnecessary.

thobe reviewed (changes requested):
It seems there was not a clean cut in the commit history of #7822 complexity wise. We traded the unused complexity of cancellation there for the complexity of "successors" here, while not getting the full benefit of not needing cancellation.

chrisvest committed:
Small WorkSync cleanups

chrisvest committed:
Make WorkSync wrap exceptions from applying work in ExecutionExceptions

Now any exception thrown during the application of work inside of the `WorkSync` now gets wrapped in an `ExecutionException`.
This makes it more similar to what `j.u.c.ExecutorService` does, except the exception is only thrown in the thread that actually does the work.
The other threads that enqueued work do not get to see if their particular unit of work caused an exception to be thrown.
The reason is two-fold: first, it's just easier to implement it this way, and second, since units of work can be combined, the notion of which unit actually caused an exception to be thrown is already quite vague.

It could be argued that all units that went into a failing work application should lead to an exception being thrown.
However, this makes the assumption that the work application is basically atomic, and that might not be the case.
In other words, the `WorkSync` code has no way of knowing how much work got applied, if any, before the exception got thrown.
Further, if the callers really do wish for all participating threads to be notified of failure through exceptions, then the `Work.combine` and `apply` methods can be made to keep track of this, by maintaining the necessary state inside the `Work` implementations themselves.

chrisvest committed:
Remove handling of InterruptedException from WorkSync

We no longer need to deal with InterruptedExceptions since we started doing our own simple try-locking using an atomic reference.
This lock implementation is uninterruptible.

chrisvest committed:
WorkSync refactorings

Extract clear responsibilities out into separate methods

chrisvest committed:
Apply code format to WorkSync code

chrisvest committed:
Move the WorkSynd done-marking out of the critical section

This allows the next batch to start processing while the current batch is waking up threads that submitted work to it.

This is incompatible with successor-based unparking, wherein a specific successor was chosen among the parked work submitters.
The job of the successor was then to unpark all of the other parked work submitters in the batch.
However, now that the done-marking has moved out of the critical section, it was then possible for threads to notice that their work was completed before they parked.
Thus, they would never wake up as successors and therefor never take the job of unparking the other waiting threads in the batch.
This was possibly a bug in the previous implementation as well.

Now, the thread that got the lock, and combined and performed the work, is now also responsible for unparking all threads in the batch.
Since this is no longer done while holding the WorkSync lock, it is perfectly okay that this thread spends the extra time to do this work.
The reason we introduced successor-based unparking was to reduce the amount of time the work-doing thread spent holding the WorkSync lock.

Quick testing shows that this also improves the performance by a few percent.

chrisvest commented:
@thobe I believe I've addressed all of your comments.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java:
```diff
@@ -19,109 +19,176 @@
 {
     private final Material material;
     private final AtomicReference<WorkUnit<Material,W>> stack;
     private final WorkUnit<Material,W> stackEnd;
```
Remind me again why `stackEnd` cannot be static? Then put that reminder into a source code comment for this field.

If it was static the sharing of it that takes place between `WorkSync` and `WorkUnit` would be simpler.

chrisvest replied:
It can be made static at the cost of some unchecked casts to get around the generics.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java:
```diff
@@ -19,109 +19,176 @@
+        return failure;
     }
 
     private WorkUnit<Material,W> reverse( WorkUnit<Material,W> batch )
```
Given that work combining should be commutative, do we really need to reverse the stack of `WorkUnit`s before combining them?

chrisvest replied:
Hmm... I guess we don't.

thobe reviewed (commented):
The code is much clearer now. So much clearer that I find myself thinking about higher order things about this code.

chrisvest committed:
Make workSync.stackEnd static because it does not need to be a per instance field

chrisvest committed:
Make WorkSync not reverse the batch prior to combining

This saves one iteration (via volatile reads) through the batch, and is allowed because the work units are commutative and associative.

chrisvest commented:
@thobe I decided to also include your last two suggestions.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java:
```diff
@@ -19,130 +19,176 @@
+ *
  * @see Work
  */
+@SuppressWarnings( {"unchecked", "NumberEquality"} )
```
Isn't the `NumberEquality` warning being erroneously raised?

chrisvest replied:
The message for the warning was relevant, so I think it's just the name of the warning. Looks like any use of reference equality will raise this warning. A bit daft, but I felt that I might as well quiet the last yellow lines while I was there.

thobe commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java:
```diff
@@ -158,7 +204,18 @@ private W combine( WorkUnit<Material,W> batch )
+                // todo Java9: Thread.onSpinWait() ?
+                Thread.yield();
+                tmp = batch.next;
+            }
```
Hmm... this is a tricky one. Here is indeed the first place where we are trying to read `WorkUnit.next` now, but it doesn't "feel" like the right place for this spin loop. Accounting for this race would feel more like it "belongs" in the `grabBatch()` method. On the other hand, the longer we can defer the check, the more likely it is we don't have to retry, meaning we should try to defer this look as long as possible - i.e. to here.

Perhaps it would be cleaner in the code to encapsulate this logic in a method, but then it would feel more natural to always use that method for accessing the field, and in `markAsDone()` we don't need to compensate for the race, since `combine()` was invoked before and thus already compensated for it. Although, the overhead of null-checks is really low, so perhaps that's acceptable.

chrisvest replied:
Having cake vs. eating cake.

thobe reviewed (commented):
Looks good. Just some minor thoughts.

thobe commented:
The build failure looks unrelated:
```
[09:44:32][Step 5/5] [INFO] --- findbugs-maven-plugin:3.0.3:check (default) @ neo4j-unsafe ---
[09:44:32][Step 5/5] [INFO] BugInstance size is 2
[09:44:32][Step 5/5] [INFO] Error size is 0
[09:44:32][Step 5/5] [INFO] Total bugs: 2
[09:44:32][Step 5/5] [INFO] Exception is caught when Exception is not thrown in org.neo4j.unsafe.impl.internal.dragons.UnsafeUtil.getSharedStringConstructorMethodHandle(MethodHandles$Lookup) [org.neo4j.unsafe.impl.internal.dragons.UnsafeUtil] At UnsafeUtil.java:[line 218] REC_CATCH_EXCEPTION
[09:44:32][Step 5/5] [INFO] org.neo4j.unsafe.impl.internal.dragons.UnsafeUtil$FreeTrace defines compareTo(UnsafeUtil$FreeTrace) and uses Object.equals() [org.neo4j.unsafe.impl.internal.dragons.UnsafeUtil$FreeTrace] At UnsafeUtil.java:[line 370] EQ_COMPARETO_USE_OBJECT_EQUALS
```
